File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    var answer = true;
    
    const map = new Map();
    
    const compareFunc = (a, b) => {
        if (a.length == b.length) {
            return a - b
        }
        else {
            return a.length - b.length
        }
    }
    
    phone_book.sort(compareFunc)
    
    for (let i = 0; i < phone_book.length; i++) {
        cur_phone_book = phone_book[i]
        
        for (let j = 0; j < cur_phone_book.length; j++) {
            const sliced_num = cur_phone_book.slice(0,j+1)
            if (map.has(sliced_num)) {
                return false
            } else {
                if (j == cur_phone_book.length - 1) {
                    map.set(cur_phone_book, 1)
                }
            }
            
        }
        
    }
    
    
    
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool solution(vector<string> phone_book) {
    // 1. 기본 사전순 정렬 (O(N log N))
    // "119", "1195524421", "97674223" 순으로 정렬됨
    sort(phone_book.begin(), phone_book.end());

    // 2. 단 한 번의 루프로 인접한 번호만 비교 (O(N))
    for (int i = 0; i < (int)phone_book.size() - 1; i++) {
        // C++20 starts_with 혹은 find() == 0 사용
        // 바로 다음 번호가 현재 번호로 시작하는지만 보면 됩니다.
        if (phone_book[i + 1].find(phone_book[i]) == 0) {
            return false;
        }
    }

    return true;
}